storage-proofs/src/zigzag_graph.rs

```diff
@@ -11,18 +11,78 @@ use crate::settings;
 
 pub const DEFAULT_EXPANSION_DEGREE: usize = 8;
 
-// Cache of node's parents.
-pub type ParentCache = HashMap<usize, Vec<usize>>;
+lazy_static! {
+    // This parents cache is currently used for the *expanded parents only*, generated
+    // by the expensive Feistel operations in the ZigZag, it doesn't contain the
+    // "base" (in the `Graph` terminology) parents, which are cheaper to compute.
+    static ref PARENT_CACHE: Arc<RwLock<HashMap<String, ParentCache>>> = Arc::new(RwLock::new(HashMap::new()));
+}
 
 // ZigZagGraph will hold two different (but related) `ParentCache`,
-// the first one for the `forward` direction and the second one
-// for the `reversed`.
-pub type TwoWayParentCache = [ParentCache; 2];
+// the first one for the `forward` direction and the second one for the `reversed`.
+#[derive(Debug, Clone)]
+pub struct ParentCache {
+    forward: Vec<Option<Vec<u32>>>,
+    reverse: Vec<Option<Vec<u32>>>,
+    // Keep the size of the cache outside the lock to be easily accessible.
+    cache_entries: u32,
+}
+
+impl ParentCache {
+    pub fn new(cache_entries: u32) -> Self {
+        ParentCache {
+            forward: vec![None; cache_entries as usize],
+            reverse: vec![None; cache_entries as usize],
+            cache_entries,
+        }
+    }
+
+    pub fn contains_forward(&self, node: u32) -> bool {
+        assert!(node < self.cache_entries);
+        self.forward[node as usize].is_some()
+    }
+
+    pub fn contains_reverse(&self, node: u32) -> bool {
+        assert!(node < self.cache_entries);
+        self.reverse[node as usize].is_some()
+    }
+
+    pub fn read_forward<F, T>(&self, node: u32, mut cb: F) -> T
+    where
+        F: FnMut(Option<&Vec<u32>>) -> T,
+    {
+        assert!(node < self.cache_entries);
+        cb(self.forward[node as usize].as_ref())
+    }
+
+    pub fn read_reverse<F, T>(&self, node: u32, mut cb: F) -> T
+    where
+        F: FnMut(Option<&Vec<u32>>) -> T,
+    {
+        assert!(node < self.cache_entries);
+        cb(self.reverse[node as usize].as_ref())
+    }
+
+    pub fn write_forward(&mut self, node: u32, parents: Vec<u32>) {
+        assert!(node < self.cache_entries);
+
+        let old_value = std::mem::replace(&mut self.forward[node as usize], Some(parents));
+
+        debug_assert_eq!(old_value, None);
```

Comment:
nit: if this is the only reason we're adding an `Option` to the cache type I would just do a read before (or some other check to see if this entry is empty) in the debug case since we'd be calling this write function only once per node.

Reply:
I think this structure is pretty good for now, we should eventually fill the cache in one go, and then we can remove the `Option` here.